managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ShadowManagedLedgerImpl.java

```diff
@@ -18,40 +18,363 @@
+        if (lastConfirmedEntry != null || currentLedger == null) {
+            return;
+        }
+        lastConfirmedEntry = new PositionImpl(currentLedger.getId(), currentLedger.getLastAddConfirmed());
+        // bypass empty ledgers, find last ledger with Message if possible.
+        while (lastConfirmedEntry.getEntryId() == -1) {
+            Map.Entry<Long, LedgerInfo> formerLedger = ledgers.lowerEntry(lastConfirmedEntry.getLedgerId());
+            if (formerLedger != null) {
+                LedgerInfo ledgerInfo = formerLedger.getValue();
+                lastConfirmedEntry = PositionImpl.get(ledgerInfo.getLedgerId(), ledgerInfo.getEntries() - 1);
+            } else {
+                break;
+            }
+        }
+    }
+
+    @Override
+    protected synchronized void internalAsyncAddEntry(OpAddEntry addOperation) {
+        if (!beforeAddEntry(addOperation)) {
+            return;
+        }
+        if (state == State.Terminated) {
+            addOperation.failed(new ManagedLedgerException.ManagedLedgerTerminatedException(
+                    "Managed ledger was already terminated"));
+            return;
+        }
+        if (state != State.LedgerOpened) {
+            addOperation.failed(new ManagedLedgerException("Managed ledger is not opened"));
+            return;
+        }
+
+        if (addOperation.getCtx() == null || !(addOperation.getCtx() instanceof Position position)) {
+            addOperation.failed(new ManagedLedgerException("Illegal addOperation context object."));
+            return;
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("[{}] Add entry into shadow ledger lh={} entries={}, pos=({},{})",
+                    name, currentLedger.getId(), currentLedgerEntries, position.getLedgerId(), position.getEntryId());
+        }
+        pendingAddEntries.add(addOperation);
+        if (position.getLedgerId() == currentLedger.getId()) {
+            // Write into lastLedger
+            addOperation.setLedger(currentLedger);
+            currentLedgerEntries = position.getEntryId();
+            currentLedgerSize += addOperation.data.readableBytes();
+            addOperation.initiateShadowWrite();
+        }
+        lastAddEntryTimeMs = System.currentTimeMillis();
+    }
+
+    /**
+     * Handle source ManagedLedgerInfo updates.
+     * Update types:
+     * 1. new ledgers.
+     * 2. old ledgers deleted.
+     * 3. old ledger offload info updated (including ledger deleted from bookie by offloader)
+     *
+     */
+    private synchronized void processSourceManagedLedgerInfo(MLDataFormats.ManagedLedgerInfo mlInfo, Stat stat) {
+
+        if (log.isDebugEnabled()) {
+            log.debug("[{}][{}] new SourceManagedLedgerInfo:{}, prevStat={},stat={}", name, sourceMLName, mlInfo,
+                    sourceLedgersStat, stat);
+        }
+
+        sourceLedgersStat = stat;
+
+        if (mlInfo.hasTerminatedPosition()) {
+            state = State.Terminated;
```

Comment:
Not needed actually, just keeping the original logic at first. 
After another thought, we should remove this, because the source topic can be terminated but the replication to the shadow topic should go on.
So comes these updates:
1. Removed Terminated state from ShadowManagedLedger.
2. Overrides `asyncTerminate` and return fail directly.